airflow/utils/log/elasticsearch_task_handler.py

```diff
@@ -0,0 +1,111 @@
+                 host='localhost:9200'):
+        """
+        :param base_log_folder: base folder to store logs locally
+        :param filename_template: log filename template
+        :param host: Elasticsearh host name
+        """
+        super(ElasticsearchTaskHandler, self).__init__(
+            base_log_folder, filename_template)
+        self.client = Elasticsearch([host])
+
+    def streaming_read(self, dag_id, task_id, execution_date,
+                  try_number, offset=None, page=0, max_line_per_page=1000):
+        """
+        Endpoint for streaming log.
+        :param dag_id: id of the dag
+        :param task_id: id of the task
+        :param execution_date: execution date in isoformat
+        :param try_number: try_number of the task instance
+        :param offset: filter log with offset strictly greater than offset
+        :param page: logs at given page
+        :param max_line_per_page: maximum number of results returned per ES query
+        :return a list of log documents
+        """
+        log_id = '-'.join([dag_id, task_id, execution_date, try_number])
+
+        s = Search(using=self.client) \
+            .query('match', log_id=log_id) \
+            .sort('offset')
+
+        # Offset is the unique key for sorting logs given log_id.
+        if offset:
+            s = s.filter('range', offset={'gt': offset})
+
+        try:
+            response = s[max_line_per_page * page:max_line_per_page].execute()
+            logs = [hit for hit in response]
+
+        except ElasticsearchException as e:
+            # Do not swallow the ES error.
+            err = "Unable to read logs from ElasticSearch: {}\n".format(str(e))
+            raise AirflowException(err)
+
+        return logs
+
+    def _read(self, ti, try_number):
+        """
+        Read all logs of given task instance and try_number from Elaticsearch.
+        :param ti: task instance object
+        :param try_number: task instance try_number to read logs from
+        :return return log messages in string format
+        """
+        dag_id = ti.dag_id
+        task_id = ti.task_id
+        execution_date = ti.execution_date.isoformat()
+
+        log_id = '-'.join([dag_id, task_id, execution_date, try_number])
+
+        # Use ES Scroll API to get all logs, since query DSL can at most
+        # return 10k results. This might take some time as sorting scroll
+        # results is very expensive. Please use streaming endpoint to boost
+        # performance.
+        s = Search(using=self.client) \
+            .query('match', log_id=log_id) \
+            .sort('offset')
+        query = s.to_dict()
+        try:
+            response = helpers.scan(self.client, query=query, preserve_order=True)
+            log = '\n'.join([hit['_source']['message'] for hit in response])
+        except ElasticsearchException as e:
+            err = "Unable to read logs from ElasticSearch: {}\n".format(str(e))
+            raise AirflowException(err)
```

Comment:
Can we create a new exception for this? I know everyone else uses AirflowException but we should really stop since the way it is used it's just like using `Exception` everywhere